clodius/tiles/bigbed.py

```diff
@@ -0,0 +1,407 @@
+            return 0
+    except TypeError:
+        return 1
+
+
+def natsorted(iterable):
+    return sorted(iterable, key=ft.cmp_to_key(natcmp))
+
+
+def get_chromsizes(bbpath):
+    """
+    TODO: replace this with negspy
+
+    Also, return NaNs from any missing chromosomes in bbi.fetch
+
+    """
+    chromsizes = bbi.chromsizes(bbpath)
+    chromosomes = natsorted(chromsizes.keys())
+    chrom_series = pd.Series(chromsizes)[chromosomes]
+    return chrom_series
+
+
+def abs2genomic(chromsizes, start_pos, end_pos):
+    abs_chrom_offsets = np.r_[0, np.cumsum(chromsizes.values)]
+    cid_lo, cid_hi = np.searchsorted(abs_chrom_offsets,
+                                     [start_pos, end_pos],
+                                     side='right') - 1
+    rel_pos_lo = start_pos - abs_chrom_offsets[cid_lo]
+    rel_pos_hi = end_pos - abs_chrom_offsets[cid_hi]
+    start = rel_pos_lo
+    for cid in range(cid_lo, cid_hi):
+        yield cid, start, chromsizes[cid]
+        start = 0
+    yield cid_hi, start, rel_pos_hi
+
+
+def tileset_info(bbpath, chromsizes=None):
+    '''
+    Get the tileset info for a bigWig file
+
+    Parameters
+    ----------
+    bbpath: string
+        The path to the bigbed file from which to retrieve data
+    chromsizes: [[chrom, size],...]
+        A list of chromosome sizes associated with this tileset.
+        Typically passed in to specify in what order data from
+        the bigbed should be returned.
+
+    Returns
+    -------
+    tileset_info: {
+                   'min_pos': [],
+                   'max_pos': [],
+                   'tile_size': 1024,
+                   'max_zoom': 7
+                  }
+    '''
+    TILE_SIZE = 1024
+
+    if chromsizes is None:
+        chromsizes = get_chromsizes(bbpath)
+        chromsizes_list = []
+
+        for chrom, size in chromsizes.iteritems():
+            chromsizes_list += [[chrom, int(size)]]
+    else:
+        chromsizes_list = chromsizes
+
+    min_tile_cover = np.ceil(
+        sum([int(c[1]) for c in chromsizes_list]) / TILE_SIZE
+    )
+    max_zoom = int(np.ceil(np.log2(min_tile_cover)))
+
+    tileset_info = {
+        'min_pos': [0],
+        'max_pos': [TILE_SIZE * 2 ** max_zoom],
+        'max_width': TILE_SIZE * 2 ** max_zoom,
+        'tile_size': TILE_SIZE,
+        'max_zoom': max_zoom,
+        'chromsizes': chromsizes_list,
+        'range_modes': range_modes
+    }
+    return tileset_info
+
+
+def fetch_data(a):
+    (
+        bbpath,
+        binsize,
+        chromsizes,
+        range_mode,
+        min_elements,
+        max_elements,
+        cid,
+        start,
+        end
+    ) = a
+    n_bins = int(np.ceil((end - start) / binsize))
+
+    try:
+        chrom = chromsizes.index[cid]
+        clen = chromsizes.values[cid]
+
+        kwargs = {"bins": n_bins, "missing": np.nan}
+        
+        fetch_factory = ft.partial(bbi.fetch_intervals, bbpath, chrom, start, end)
+
+        if range_mode == 'significant':
```

Comment:
`significant` vs. `significance`?

Reply:
And since there's no difference between including that option and not including it, maybe it's better to leave it out until it causes different behavior?

Reply:
Unless you have a strong objection to this, I suggest leaving in a descriptive name for the range mode, as there may be other approaches used for querying and subsetting ranges that have not been considered, and this helps document what the current approach is (even if it is the only option I wrote, and the default one).

Reply:
I agree that keeping it makes sense. Could you document how the `significance` filter is downsampling the intervals in a tile? If it's not obvious where to put it, you could put it in a module-level docstring for now.

I also don't know if bigBeds are guaranteed to have a column 5 or that it be an integer `score` column. Maybe there needs to be a fallback to treating such files as having score = 1 for uniform downsampling.